go/src/koding/socialapi/cache.go

```diff
@@ -0,0 +1,165 @@
+package socialapi
+
+import (
+	"errors"
+	"sync"
+)
+
+// ErrSessionNotFound is returned by Storage.Get
+// when requested session was not found.
+var ErrSessionNotFound = errors.New("session not found")
+
+// Storage is an interface for external session storage.
+type Storage interface {
+	// Get gets a session.
+	//
+	// It returns ErrSessionNotFound if requested
+	// session was not cached.
+	Get(*Session) error
+
+	// Set sets a session.
+	//
+	// It overwrites any session that may exist
+	// already.
+	Set(*Session) error
+
+	// Delete deletes a session.
+	//
+	// If requested session does not exist,
+	// the method is a nop.
+	Delete(*Session) error
+}
+
+// SessionCache is a wrapper for AuthFunc that caches
+// sessions and invalidates them when requested.
+type SessionCache struct {
+	// AuthFunc is a mean to obtain session information.
+	//
+	// It is only called when a session was requested
+	// for a user/team pair that is not cached yet.
+	// Or when we were explicitly asked to request
+	// new session with Refresh set to true.
+	AuthFunc AuthFunc
+
+	// Storage is an external storage for storing session.
+	//
+	// If nil, in-memory map is going to be used instead.
+	Storage Storage
+
+	cacheMu sync.RWMutex
+	cache   map[string]*Session
```

Comment:
i believe you wrote these in frustration :) the Rafal i know does not complain instead fixes them whenever he see a problem. 

Other than that i would strongly suggest you to use LRU instead of simple memory cache, since the talk all started from leaking memory not the trivial implementation itself. In the previous-previous talk we both were OK with a simple map